# Assignment 5
### Do all four questions.

Mason L Earp

**1.** Let's review some basic matrix multiplication. When you have an $M \times N$ matrix $A$ with $M$ rows and $N$ columns, 
$$
A= \left[ \begin{array}{cccc} a_{11} & a_{12} & ... & a_{1N} \\
a_{21} & a_{22} & ... & a_{2N} \\
\vdots & \vdots & ... & \vdots \\
a_{M1} & a_{M2} & ... & a_{MN} 
\end{array} \right],
$$
and you right-multiply it by a vector
$$
x = \left[ \begin{array}{c} x_1 \\ x_2 \\ \vdots \\ x_N 
\end{array} \right],
$$
you get
$$
Ax = \left[ \begin{array}{c} \sum_{i=1}^N a_{1i} x_i \\ \sum_{i=1}^N a_{2i} x_i \\ \vdots \\ \sum_{i=1}^N a_{Mi} x_i 
\end{array} \right].
$$
This is just "matrix row times column vector" element-by-element, stacking the results into a new vector.

For this to make sense, $N$ must be the same for the matrix and the vector, but $M$ can be different from $N$. 

Let's play with some NumPy to see this. First we'll define a matrix $A$:

In [57]:
import numpy as np

A = np.array([ [1,2,3],
              [4,5,6],
              [7,8,9]])
A

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

a. Multiply $A$ times each of the following vectors using the @ operator. Explain which part of the $A$ matrix gets selected and explain why, using the definition of matrix multiplication. 

In [58]:
e_1 = np.array([1,0,0])
e_2 = np.array([0,1,0])
e_3 = np.array([0,0,1])

In [59]:
print(e_1)

[1 0 0]


In [60]:
q1 = A@e_1
q2 = A@e_2
q3 = A@e_3

In [61]:
print(q1)

[1 4 7]


The operation `A@e_1` is effectivley getting the first column of the A matrix. Another way to think about this is that `A@e_1` is getting the first value from each row in the A Matrix. 

The Matrix Math is that each row of A is multiplied by the vector e_1[1,0,0] \
$ROW 1 is (1*1) + (2*0) +(3*0) = 1$\
$ROW 2 is (4*1) + (5*0) +(6*0) = 4$\
$ROW 3 is (7*1) + (8*0) +(9*0) = 7$

In [62]:
print(q2)

[2 5 8]


The operation `A@e_2` is effectivley getting the middle column of the A matrix. Another way to think about this is that `A@e_2` is getting the middle value from each row in the A Matrix. 

The Matrix Math is that each row of A is multiplied by the vector e_1[0,1,0] \
$ROW 1 is (1*0) + (2*1) +(3*0) = 2$\
$ROW 2 is (4*0) + (5*1) +(6*0) = 5$\
$ROW 3 is (7*0) + (8*1) +(9*0) = 8$

In [63]:
print(q3)

[3 6 9]


The operation `A@e_3` is effectivley getting the third column of the A matrix. Another way to think about this is that `A@e_3` is getting the third value from each row in the A Matrix. 

The Matrix Math is that each row of A is multiplied by the vector e_1[0,0,1] \
$ROW 1 is (1*0) + (2*0) +(3*1) = 3$\
$ROW 2 is (4*0) + (5*0) +(6*1) = 6$\
$ROW 3 is (7*0) + (8*0) +(9*1) = 9$

In [64]:
#To test the above assumptions im going to make a 4 row Matrix (B) with the added row of [10, 11, 12]
B = np.array([ [1,2,3],
              [4,5,6],
              [7,8,9],
              [10,11,12]])
print(B)
#Then multiply B by e_1, e_2, e_3
#Expected output of B@e_1 is [1, 4, 7, 10]
q4 = B@e_1
print(q4)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]
[ 1  4  7 10]


b. Now multiply $A$ times $u = (1,1,1)$. Explain the logic of the result with the definition of matrix multiplication.

In [65]:
u = np.ones(3)
u

array([1., 1., 1.])

In [66]:
q = A@u
q

array([ 6., 15., 24.])

The result is the sum of the rows, from matrix A. So row 1 is 6=1+2+3, Row 2 is 15=4+5+6, Row 3 is 24=7+8+9.
 
The Matrix Math is that each row of A is multiplied by the vector u[1,1,1] With the rows summed \
$ROW 1 is (1*1) + (2*1) +(3*1) = 6$\
$ROW 2 is (4*1) + (5*1) +(6*1) = 15$\
$ROW 3 is (7*1) + (8*1) +(9*1) = 24$

c. Whenever a matrix has 1's on the diagonal and zeros everywhere else, we call it an **identity matrix**. What happens when you multiple $A$ times $x$ below? What happens when you multiple an identity matrix times any vector? Explain your result with the definition of matrix multiplication.

In [67]:
A = np.array([ [1,0,0],
              [0,1,0],
              [0,0,1]])
x = np.array([-2,4,11])
q1 = x@A
q1


array([-2,  4, 11])

Multiplying an identity matrix by any vecotr will return the original vector. For the Array x = [-2,4,11] The identity array A will mutiply the ith element by the jth column.

The Matrix Math is the row in `x` is multiplied by each column in `A` \
$(-2*1) + (4*0) +(11*0) = -2$\
$(-2*0) + (4*1) +(11*0) = 4$\
$(-2*0) + (4*0) +(11*1) = 11$


d. What if every row and column sum to 1, but the 1's are no longer on the diagonal? Multiple $A$ times $X$ below and explain the result. Create another matrix whose rows and columns sum to 1, but is not an identity matrix, and show how it permutes the values of $x$. 

In [68]:
A = np.array([ [0,0,1],
              [1,0,0],
              [0,1,0]])
x = np.array([-2,4,11])
q = A@x
q

array([11, -2,  4])

By moving where the 1's and 0's are in the new matrix we are reordering where the values are from the original matrix.\
Matrix A =  \
[0,0,1],\
              [1,0,0],\
              [0,1,0]\
              Effectively moves the third component to the first, the first component to the second, and the second component to the third. \
$(-2*0) + (4*0) +(11*1) = 11$\
$(-2*1) + (4*0) +(11*0) = -2$\
$(-2*0) + (4*1) +(11*0) = 4$

e. The next matrix $A$ could be a Markov transition matrix: Its columns sum to 1, and each entry $a_{ij}$ can be interpreted as the proportion of observations who moved from state $j$ to state $i$. Multiply $A$ by each of the vectors $e_1$, $e_2$, and $e_3$, and explain your results.

In [69]:
rng = np.random.default_rng(100)
A = rng.random((3,3)) # Generate a random 3X3 matrix
sums = np.sum(A,axis=0) # Column sums
A = A/sums # Normalize the columns so they sum to 1
print(A)

[[0.50052958 0.24049286 0.18358131]
 [0.02574731 0.39251588 0.37907577]
 [0.47372311 0.36699127 0.43734292]]


In [70]:
e_1 = np.array([1,0,0])
e_2 = np.array([0,1,0])
e_3 = np.array([0,0,1])

e1 = A@e_1
e2 = A@e_2
e3 = A@e_3

print(f"Here's A@e_1: ", e1)
print(f"Here's A@e_2: ", e2)
print(f"Here's A@e_3: ", e3)

Here's A@e_1:  [0.50052958 0.02574731 0.47372311]
Here's A@e_2:  [0.24049286 0.39251588 0.36699127]
Here's A@e_3:  [0.18358131 0.37907577 0.43734292]


The resulting vectors are the columns from the randomized matrix, each one adds to 1.\
Each multiplication shows the proabability that an observation moved to a different state:\
Here's A@e_1:  \
0.50052958 stayed in state 1, 1->1\
0.02574731 transitioned to state 2, 1->2\
0.47372311 transitioned to state 3, 1->3

Here's A@e_2:  \
0.24049286 transitioned to state 1, 2->1\
0.39251588 stayed in state 2, 2->2\
0.36699127 transitioned to state 3, 3->3

Here's A@e_3:  \
0.18358131 transitioned to state 1, 3->1\
0.37907577 transitioned to state 2, 3->2\
0.43734292 stayed in state 3, 3->3

f. For each of the vectors $e_1, e_2, e_3$, multiple $A$ times that vector 5 times. What answer do you get for each starting vector? Describe the behavior you observe.

In [71]:
rng = np.random.default_rng(100)
A = rng.random((3,3)) # Generate a random 3X3 matrix
sums = np.sum(A,axis=0) # Column sums
A = A/sums # Normalize the columns so they sum to 1

e_1 = np.array([1,0,0])
e_2 = np.array([0,1,0])
e_3 = np.array([0,0,1])

e = e_1
for i in range(5):
    e = A @ e
print(f"Here's A@e_1 5 times: ", e)



Here's A@e_1 5 times:  [0.29266551 0.27862515 0.42870935]


In [72]:
e = e_2
for i in range(5):
    e = A @ e
print(f"Here's A@e_2 5 times: ", e)

Here's A@e_2 5 times:  [0.29197422 0.27979983 0.42822595]


In [73]:
e = e_3
for i in range(5):
    e = A @ e
print(f"Here's A@e_3 5 times: ", e)

Here's A@e_3 5 times:  [0.29171646 0.2802254  0.42805814]


It appears that the transitioning between states is converging. I would assume that after thousands of iterations we would reach a steady state that equals ~.30, ~.30, ~.40 \
The probability of transitioning to 1 = [0.29266551,0.29197422,0.29171646]\
The probability of transitioning to 2 = [0.27862515,0.27862515,0.27862515]\
The probability of transitioning to 3 = [0.42870935,0.42822595,0.42805814]

*2.* Let's consider a simple Markov transition matrix over two states:
$$
T = \left[ \begin{array}{cc} p_{1\leftarrow 1} &  p_{1\leftarrow 2} \\
p_{2 \leftarrow 1} & p_{2 \leftarrow 2} \end{array}\right] 
$$
The arrows help visualize the transition a bit: This is the same index notation as usual, $p_{ij}$, but writing it $p_{i \leftarrow j}$ emphasizes that it's the proportion of times that state $j$ transitions to state $i$. Below, $T$ is given by
$$
T = \left[ \begin{array}{cc} .25 & .5 \\
.75 & .5 \end{array}\right].
$$

- Start in state 1, at the initial condition $[1,0]$. Multiply that vector by $T$. Write out the result in terms of the formula and compute the result in a code chunk below. What is this object you're looking at, in terms of proportions and transitions?
- Multiple by $T$ again. What do you get? This isn't a column of $T$. Explain in words what it is. (Hint: A forecast of what in what period?)
- Keep multiplying the current vector of outcomes by $T$. When does it start to settle down without changing further?
- Do the above analysis again, starting from the initial condition $[0,1]$. Do you get a different result?
- The take-away is that, in the long run, these chains settle down into the long-run proportions, and the sensitivity on initial conditions vanishes. 


In [74]:
init_condition = np.array([1,0])
print(init_condition)

[1 0]


In [75]:
T = np.array([[ 1/4, 1/2],
                 [ 3/4, 1/2 ]])
print(T)

[[0.25 0.5 ]
 [0.75 0.5 ]]


In [76]:
p1 = T @ init_condition
p1

array([0.25, 0.75])

The resulting array `p1` is the probability that the states transitioned from the 1 state. The first value 0.25 is the proabablity that the state did not change or went from $1 \rightarrow 1$ . The second value is the probablity that the state went from $1 \rightarrow 2$ .
$$
p1 = \left[ \begin{array}{cc} p_{1 \rightarrow 1} (0.25)(1) &+& p_{2 \rightarrow 1}(0.5)(0) \\
p_{1 \rightarrow 2} (0.75)(1) &+& p_{2 \rightarrow 2}(0.5)(0) \end{array}\right]
$$

In [77]:
p2 = T @ p1
p2

array([0.4375, 0.5625])

p2 is the resulting probability of being is state 1 or 2 after 2 steps. Initially we were in [1,0] then moved to [0.25,0.75] after 1 more time step we are in [0.4375,0.5635]
That is the the probabilty that we are in the 1 state is 0.4375 and the probability of begin in the 2 state is 0.5625.
$$
p2 = \left[ \begin{array}{cc} (1/4)(0.25) &+& (1/2)(0.75) \\
(3/4)(0.25) &+& (1/2)(0.75) \end{array}\right]
$$
$$
p2 = \left[ \begin{array}{cc} (0.0625) &+& (0.375) \\
(0.1875) &+& (0.375) \end{array}\right]
$$
$$
p2 = \left[ \begin{array}{cc} (0.4375) \\
(0.5625) \end{array}\right]
$$

In [78]:
T = np.array([[ 1/4, 1/2],
                 [ 3/4, 1/2 ]])
p = [1,0]
sim = []
for i in range(15):
    p = T@p
    sim.append(p)
sim

[array([0.25, 0.75]),
 array([0.4375, 0.5625]),
 array([0.390625, 0.609375]),
 array([0.40234375, 0.59765625]),
 array([0.39941406, 0.60058594]),
 array([0.40014648, 0.59985352]),
 array([0.39996338, 0.60003662]),
 array([0.40000916, 0.59999084]),
 array([0.39999771, 0.60000229]),
 array([0.40000057, 0.59999943]),
 array([0.39999986, 0.60000014]),
 array([0.40000004, 0.59999996]),
 array([0.39999999, 0.60000001]),
 array([0.4, 0.6]),
 array([0.4, 0.6])]

When the initial condition is [1,0] after 15 iterations the states stabilize to [0.4, 0.6].

In [79]:
T = np.array([[ 1/4, 1/2],
                 [ 3/4, 1/2 ]])
p = [0,1]
sim = []
for i in range(15):
    p = T@p
    sim.append(p)
sim

[array([0.5, 0.5]),
 array([0.375, 0.625]),
 array([0.40625, 0.59375]),
 array([0.3984375, 0.6015625]),
 array([0.40039062, 0.59960938]),
 array([0.39990234, 0.60009766]),
 array([0.40002441, 0.59997559]),
 array([0.3999939, 0.6000061]),
 array([0.40000153, 0.59999847]),
 array([0.39999962, 0.60000038]),
 array([0.4000001, 0.5999999]),
 array([0.39999998, 0.60000002]),
 array([0.40000001, 0.59999999]),
 array([0.4, 0.6]),
 array([0.4, 0.6])]

When the initial condition is [0,1], after 15 iterations the states stabilize to [0.4,0.6]. The same as when starting at [1,0].

3. Weather data

- Load the `cville_weather.csv` data. This includes data from Jan 4, 2024 to Feb 2, 2025. Are there any missing data issues?
- Based on the precipitation variable, `PRCP`, make a new variable called `rain` that takes the value 1 if `PRCP`>0 and 0 otherwise.
- Build a two-state Markov chain over the states 0 and 1 for the `rain` variable. 
- For your chain from c, how likely is it to rain if it was rainy yesterday? How likely is it to rain if it was clear yesterday?
- Starting from a clear day, forecast the distribution. How quickly does it converge to a fixed result? What if you start from a rainy day?
- Conditional on being rainy, plot a KDE of the `PRCP` variable.
- Describe one way of making your model better for forecasting and simulation the weather.

Congratulations, you now are a non-parametric meteorologist!

In [105]:
import pandas as pd
weather = pd.read_csv("cville_weather.csv")
weather = weather.sort_values(by='DATE')
weather

STATION                            NAME        DATE  DAPR  \
0    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-04   NaN   
1    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-07   NaN   
271  US1VAAB0010    CHARLOTTESVILLE 8.4 W, VA US  2024-01-08   NaN   
2    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-09   NaN   
3    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-10   NaN   
..           ...                             ...         ...   ...   
266  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-18   NaN   
267  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-21   NaN   
268  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-25   NaN   
269  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-27   NaN   
270  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-28   NaN   

    DAPR_ATTRIBUTES  MDPR MDPR_ATTRIBUTES  PRCP PRCP_ATTRIBUTES  SNOW  \
0               NaN   NaN             NaN  0.03             ,,N   NaN   
1               NaN   NaN             NaN  1.08             ,,N   NaN   
271             NaN   NaN             NaN  0.00             ,,N   0.0   
2               NaN   NaN             NaN  0.24             ,,N   NaN   
3               NaN   NaN             NaN  3.00             ,,N   NaN   
..              ...   ...             ...   ...             ...   ...   
266             NaN   NaN             NaN  0.04             ,,N   NaN   
267             NaN   NaN             NaN  0.38             ,,N   NaN   
268             NaN   NaN             NaN  0.29             ,,N   NaN   
269             NaN   NaN             NaN  0.24             ,,N   NaN   
270             NaN   NaN             NaN  0.60             ,,N   NaN   

    SNOW_ATTRIBUTES  SNWD SNWD_ATTRIBUTES  
0               NaN   NaN             NaN  
1               NaN   NaN             NaN  
271             ,,N   NaN             NaN  
2               NaN   NaN             NaN  
3               NaN   NaN             NaN  
..              ...   ...             ...  
266             NaN   NaN             NaN  
267             NaN   NaN             NaN  
268             NaN   NaN             NaN  
269             NaN   NaN             NaN  
270             NaN   NaN             NaN  

[411 rows x 13 columns]

In [106]:
weather.isna().sum()


STATION              0
NAME                 0
DATE                 0
DAPR               399
DAPR_ATTRIBUTES    399
MDPR               399
MDPR_ATTRIBUTES    399
PRCP                12
PRCP_ATTRIBUTES     12
SNOW               188
SNOW_ATTRIBUTES    188
SNWD               410
SNWD_ATTRIBUTES    410
dtype: int64

Yes there are many there are only 411 rows and some rows have 410 missing values. But for our data i think we can just limit the frames to the date and the PRCP and remove any NaN rows from PRCP.

In [112]:
#Drop NaN values from weather in PRCP
cleaned_weather = weather.dropna(subset=['PRCP']).copy()
cleaned_weather = cleaned_weather[['STATION', 'NAME', 'DATE','PRCP']]

In [113]:
cleaned_weather['rain'] = cleaned_weather['PRCP'] > 0
cleaned_weather['rain'] = cleaned_weather['rain'].astype(int)
cleaned_weather

STATION                            NAME        DATE  PRCP  rain
0    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-04  0.03     1
1    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-07  1.08     1
271  US1VAAB0010    CHARLOTTESVILLE 8.4 W, VA US  2024-01-08  0.00     0
2    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-09  0.24     1
3    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-10  3.00     1
..           ...                             ...         ...   ...   ...
266  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-18  0.04     1
267  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-21  0.38     1
268  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-25  0.29     1
269  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-27  0.24     1
270  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-28  0.60     1

[399 rows x 5 columns]

Okay so now we have a condensed dataframe with values for rain as 1 or 0. Our next step is to count the 4 transitions $0 \rightarrow 0$, $0 \rightarrow 1$, $1 \rightarrow 0$, $1 \rightarrow 1$. Where 0 is Sunny and 1 is Rainy

In [117]:
#going from state to state we need a next state column
cleaned_weather['next_state'] = cleaned_weather['rain'].shift(-1)
cleaned_weather = cleaned_weather.dropna()
# Ensure the states are integers (0 or 1)
cleaned_weather['next_state'] = cleaned_weather['next_state'].astype(int)
cleaned_weather

C:\Users\mason\AppData\Local\Temp\ipykernel_13928\1928348905.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaned_weather['next_state'] = cleaned_weather['rain'].shift(-1)


STATION                            NAME        DATE  PRCP  rain  \
0    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-04  0.03     1   
1    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-07  1.08     1   
271  US1VAAB0010    CHARLOTTESVILLE 8.4 W, VA US  2024-01-08  0.00     0   
2    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-09  0.24     1   
3    US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2024-01-10  3.00     1   
..           ...                             ...         ...   ...   ...   
264  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-07  0.04     1   
265  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-17  0.41     1   
266  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-18  0.04     1   
267  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-21  0.38     1   
268  US1VACRC002  CHARLOTTESVILLE 0.5 NNE, VA US  2025-09-25  0.29     1   

     next_state  
0             1  
1             0  
271           1  
2             1  
3             1  
..          ...  
264           1  
265           1  
266           1  
267           1  
268           1  

[397 rows x 6 columns]

In [118]:
#count the transitions
transition_counts = pd.crosstab(
    cleaned_weather['rain'], 
    cleaned_weather['next_state'],
    rownames=['From State (Today)'],
    colnames=['To State (Tomorrow)']
)
transition_counts

To State (Tomorrow)    0    1
From State (Today)           
0                    172   48
1                     48  129

In [126]:
#Normalize the Transitions
T = transition_counts.apply(lambda x: x / x.sum(), axis=0)
T_matrix = T.to_numpy()
T.index = ['To 0', 'To 1']
T.columns = ['From 0', 'From 1']
print(T_matrix)
print(T)

[[0.78181818 0.27118644]
 [0.21818182 0.72881356]]
        From 0    From 1
To 0  0.781818  0.271186
To 1  0.218182  0.728814


How likely is it to rain if it was rainy yesterday $(1 \rightarrow 1)$ How likely is it to rain if it was clear yesterday $(0 \rightarrow 1)$?
$1 \rightarrow 1 = 72.88%$ that it will rain if it was rainy yesterday . $0 \rightarrow 1 = 21.82%$ that it will rain if it was clear yesterday.

4. Taxicab trajectories: Using the pickled taxicab data, we want to complete the exercise from class.

- For the taxicab trajectory data, determine your state space and clean your sequences of cab rides.
- Compute the transition matrix for the taxicab data between neighborhoods in Manhattan. Plot it in a heat map. What are the most common routes?
- Explain why taxicabs are most likely order 1, and not 2 or more.
- Starting at Hell's Kitchen, create a sequence of forecasts of where the cab is likely to be in 2, 3, 5, and 10 trips
- Starting at any neighborhood, iterate your forecast until it is no longer changing very much. Where do cabs spend most of their time working in Manhattan?

Going through the majority of the taxicab data in Class

In [81]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import pickle

In [82]:
with open('taxicab.pkl','rb') as f:
    data = pickle.load(f)
print(len(data))
data[0]

1000


0     Outside Manhattan
0     Outside Manhattan
0     Outside Manhattan
0     Outside Manhattan
0     Outside Manhattan
            ...        
29                 SoHo
29                 SoHo
13    Greenwich Village
3               Chelsea
3               Chelsea
Name: nbhd, Length: 26026, dtype: object